test/runtests.jl

```diff
@@ -513,6 +513,81 @@ const verb = VERSION.major == 1 && VERSION.minor < 11
         @test_throws ArgumentError Data.Floats(;minimum=2.0, maximum=1.0)
     end
 
+    # Tests the properties of the enocding used to represent floating point numbers
+    @testset "Floating point encoding" begin
+        @testset for T in (Float16, Float32, Float64)
+
+            iT = FloatEncoding.uint(T)
+            # These invariants are ported from Hypothesis
+            @testset "Exponent encoding" begin
+                # The highest value here is 0x7ff (2047) with Int64,
+                # so we can just exhaustively test all 2048 possibilities
+                exponents = zero(iT):FloatEncoding.max_exponent(T)
+
+                # Round tripping
+                @test all(exponents) do e
+                    FloatEncoding.decode_exponent(FloatEncoding.encode_exponent(e)) == e
+                end
+
+                @test all(exponents) do e
+                    FloatEncoding.encode_exponent(FloatEncoding.decode_exponent(e)) == e
+                end
+            end
+
+            function roundtrip_encoding(f)
+                encoded = FloatEncoding.float_to_lex(f)
+                decoded = FloatEncoding.lex_to_float(T, encoded)
+                reinterpret(iT, decoded) == reinterpret(iT, f)
```

Comment:
```suggestion
                reinterpret(iT, decoded) === reinterpret(iT, f)
```

This should fix the failure on nightly, since

```
julia> NaN == NaN
false

julia> NaN === NaN
true
```

The latter also catches different NaN-bitpatterns, which shouldn't happen (i.e. roundtripping should preserve the exact bitpattern, if I'm not mistaken).

Reply:
You're right, I missed that!

```
julia> @check roundtrip_encoding(Data.Floats{T}(; minimum=nextfloat(zero(T))))
Found counterexample
  Context: roundtrip_encoding

  Arguments:
      arg_1::Float16 = NaN16

  Events:
    input
        "1111111000000001"
    UNLABELED_EVENT_1
        "1111111000000001"
    UNLABELED_EVENT_2
        "0111111000000001"
```

Is `-Inf` also affected? If so, we can set `nans=false, infs=false` and just test those explicitly.

Reply:
I think Infs should be okay because `-Inf < 0`. This issue with `-NaN` I guess is that it can't compare with anything so it isn't caught by clamping; that's probably reasonable since semantically the sign doesn't matter.

Hopefully the commit I just pushed should handle testing all this.

Reply:
I had forgotten about the events API by the way. That's really useful!

Reply:
One thing that I'm wondering about is whether a `minimum` that's >= 0 should imply that the signbit is unset, even for `NaN` :thinking:

Reply:
That might be somewhat confusing given the semantics of NaN. It might be better to just add a note to the docstring that the possibility can generate any NaN bit pattern. Maybe with a flag or an example of how to convert all nans to the "canonical" nan.